src/GLVertexBuffer.hpp

```diff
@@ -0,0 +1,69 @@
+#ifndef GLVERTEXBUFFER_H
+#define GLVERTEXBUFFER_H
+
+#include <vector>
+#include <utility>
+
+#include "utils.hpp"
+
+class GLVertexBuffer {
+  public:
+    GLVertexBuffer(std::vector<std::pair<int, int>> dims)
+      : attributeDimensions(std::move(dims))
+    {
+      vertexAttribFirstPtr = vertexAttribNextPtr;             
+      vertexAttribNextPtr += attributeDimensions.size();
+
+      for (const auto& dim : attributeDimensions) {
+        strideLength += dim.first * dim.second;
+      }
+    }
+
+    ~GLVertexBuffer() {}
+    
+    template <typename NumericType, size_t BufferSize>
+    void generateBuffersFromData(NumericType data[BufferSize]);
+
+    void enableBuffers() { glBindVertexArray(VAO); }
+
+    int getNumVertices() { return numVertices; }
+
+  private:
+    std::vector<std::pair<int, int>> attributeDimensions;
+    int vertexAttribFirstPtr;
+    int numVertices = 0;
+    int strideLength = 0;   // independent of numeric type size
+    unsigned int VBO, VAO;
+
+    static int vertexAttribNextPtr;
+};
+
+template <typename NumericType, size_t BufferSize>
+void GLVertexBuffer::generateBuffersFromData(NumericType data[BufferSize]) {
+  numVertices = BufferSize / strideLength;
+
+  glGenBuffers(1, &VBO);
+  glGenVertexArrays(1, &VAO);
+
+  glBindBuffer(GL_ARRAY_BUFFER, VBO);
+  glBufferData(GL_ARRAY_BUFFER, BufferSize*sizeof(float), data, GL_STATIC_DRAW);
```

Comment:
Do we want this to be `sizeof(int)` if `NumericType` is `GLint`?

Reply:
Thanks for catching this. It should be `NumericType` instead.